pkg/kubelet/cm/topologymanager/topology_manager.go

```diff
@@ -47,7 +65,216 @@ type Store interface {
+			}
+		}
+
+		// Merge the affinities using a bitwise-and operation.
+		mergedAffinity, _ := socketmask.NewSocketMask()
+		mergedAffinity.Fill()
+		mergedAffinity.And(socketAffinities...)
+
+		// Build a mergedHintfrom the merged affinity mask, indicating if an
+		// preferred allocation was used to generate the affinity mask or not.
+		mergedHint := TopologyHint{mergedAffinity, preferred}
+
+		// Only consider mergedHints that result in a SocketAffinity > 0 to
+		// replace the current bestHint.
+		if mergedHint.SocketAffinity.Count() == 0 {
+			return
+		}
+
+		// If the current bestHint is non-preferred and the new mergedHint is
+		// preferred, always choose the preferred hint over the non-preferred one.
+		if mergedHint.Preferred && !bestHint.Preferred {
+			bestHint = mergedHint
+			return
+		}
+
+		// If the current bestHint is preferred and the new mergedHint is
+		// non-preferred, never update bestHint, regardless of mergedHint's
+		// narowness.
+		if !mergedHint.Preferred && bestHint.Preferred {
+			return
+		}
+
+		// If mergedHint and bestHint has the same preference, only consider
+		// mergedHints that have a narrower SocketAffinity than the
+		// SocketAffinity in the current bestHint.
+		if !mergedHint.SocketAffinity.IsNarrowerThan(bestHint.SocketAffinity) {
+			return
+		}
+
+		// In all other cases, update bestHint to the current mergedHint
+		bestHint = mergedHint
+	})
+
+	klog.Infof("[topologymanager] ContainerTopologyHint: %v", bestHint)
+
+	return bestHint
+}
+
+func (m *manager) AddHintProvider(h HintProvider) {
+	m.hintProviders = append(m.hintProviders, h)
+}
+
+func (m *manager) AddContainer(pod *v1.Pod, containerID string) error {
+	m.podMap[containerID] = string(pod.UID)
+	return nil
+}
+
+func (m *manager) RemoveContainer(containerID string) error {
+	podUIDString := m.podMap[containerID]
+	delete(m.podTopologyHints, podUIDString)
+	delete(m.podMap, containerID)
+	klog.Infof("[topologymanager] RemoveContainer - Container ID: %v podTopologyHints: %v", containerID, m.podTopologyHints)
+	return nil
+}
+
+func (m *manager) Admit(attrs *lifecycle.PodAdmitAttributes) lifecycle.PodAdmitResult {
+	klog.Infof("[topologymanager] Topology Admit Handler")
+	pod := attrs.Pod
+	c := make(map[string]TopologyHint)
+	klog.Infof("[topologymanager] Pod QoS Level: %v", pod.Status.QOSClass)
+
+	if pod.Status.QOSClass == v1.PodQOSGuaranteed {
+		for _, container := range append(pod.Spec.InitContainers, pod.Spec.Containers...) {
+			result := m.calculateAffinity(*pod, container)
+			admitPod := m.policy.CanAdmitPodResult(result.Preferred)
+			if admitPod.Admit == false {
+				return admitPod
+			}
+			c[container.Name] = result
+		}
+		m.podTopologyHints[string(pod.UID)] = c
+		klog.Infof("[topologymanager] Topology Affinity for Pod: %v are %v", pod.UID, m.podTopologyHints[string(pod.UID)])
+
+	} else {
+		klog.Infof("[topologymanager] Topology Manager only affinitises Guaranteed pods.")
```

Comment:
Why is that? I can have 4 devices on the socket 0 and another 4 on socket 1. Devices can be linked as in example of Nvidia with NVlink. Pod can request 2 devices and it is expected that those two will be allocated from same socket, regardless of other resources that affects PodQOS* classification.

Reply:
I agree. This probably shouldn't be a constraint in the `TopologyManager`, and rather each `HintProvider` should decide this for themselves. I think it is OK, for the initial, `alpha` implementation though. We can revisit it before this feature goes beta.

Reply:
Actually. I take it back. In order for the `CPUManager` to allocate CPUs in an exclusive manner, the pod has to be in the Guaranteed QoS tier. The hints are useless if the `CPUManager` can't actually act on them, and it can't if the Pod is not Guaranteed.

Reply:
@klueska this is only in case if the CPU manager policy is `static`. If it is `none` - there is no need of Pod to be Guaranteed. If static policy gets improved in next years, it is also not going to be a problem. There is no big issues to dynamically adjust cpuset for pods to be aligned to the socket even in shared pool, it is just limitation on currently implemented `static` policy.

Reply:
At least right now if it is `none`, then the hints don't provide any useful information. Nothing can be acted on by the hints. I do get your point though. Other `CPUManager` policies could be created in the future that _could_ act on the hints even if the QoS tier isn't Guaranteed, so we shouldn't artificially limit the `TopologyManager` to accept hints in this case alone. I still think it's fine for this `alpha`, but I agree we should revisit.

Reply:
@klueska or @kad, mind making a ticket about this?

Reply:
I am going to go ahead and resolve this conversation, and we can revisit this discussion after alpha. Thanks!